File: server/src/storage/backup-manifest/index-manifest.model.ts
import { from, Observable } from 'rxjs';

import { hashBuffer } from '../../utils/lodash.utils';
import { IndexFileEntry } from './index-file-entry.model';
import {
  EntryType,
  FileManifest,
  FileManifestJournalEntryAddOrModify,
  FileManifestJournalEntryRemove,
} from './object-proto.model';

/**
 * List of manifest of a manifest file (and the journal).
 */
export class IndexManifest {
  private files = new Map<string, IndexFileEntry>();

  /**
   * Number of element in the index manifest.
   */
  get indexSize(): number {
    return this.files.size;
  }

  process(journalEntry: FileManifestJournalEntryAddOrModify | FileManifestJournalEntryRemove): void {
    if (journalEntry.type === EntryType.REMOVE) {
      this.remove(journalEntry.path);
    } else {
      this.add(journalEntry.manifest);
    }
  }

  add(manifest: FileManifest): void {
    const key = hashBuffer(manifest.path);
    if (!this.files.has(key)) {
      this.files.set(key, new IndexFileEntry(manifest));
    }
  }

  remove(filePath: Buffer): void {
    const key = hashBuffer(filePath);
    this.files.delete(key);
  }

  mark(index: IndexFileEntry): void {
    index.markViewed = true;
  }

  /**
   * Read all the file in the walk;
   * @param filter
   */
  walk(): Observable<IndexFileEntry> {
    return from(this.files.values());
  }

  /**
   * Get the index entry for the given file path;
   */
  getEntry(filePath: Buffer): IndexFileEntry | undefined {
    const key = hashBuffer(filePath);
    return this.files.get(key);
  }
}

File: shared-rs/src/index_manifest_model.rs
use std::collections::HashMap;

use crate::woodstock::{EntryType, FileManifest, FileManifestJournalEntry};

pub struct IndexFileEntry {
  pub mark_viewed: bool,
  pub manifest: FileManifest,
}

impl IndexFileEntry {
  pub fn path(&self) -> &Vec<u8> {
    &self.manifest.path
  }
}

pub struct IndexManifest {
  files: HashMap<Vec<u8>, IndexFileEntry>,
}

impl IndexManifest {
  pub fn new() -> Self {
    Self {
      files: HashMap::new(),
    }
  }

  pub fn apply(&mut self, journal_entry: FileManifestJournalEntry) {
    if journal_entry.r#type() == EntryType::Remove {
      if let Some(manifest) = journal_entry.manifest {
        self.remove(&manifest.path);
      }
    } else {
      if let Some(manifest) = journal_entry.manifest {
        self.add(manifest);
      }
    }
  }

  pub fn walk(&mut self) -> impl Iterator<Item = &IndexFileEntry> {
    self.files.values()
  }

  pub fn indexSize(&self) -> usize {
    self.files.len()
  }

  pub fn add(&mut self, manifest: FileManifest) {
    let key = manifest.path.clone();
    self.files.insert(
      key,
      IndexFileEntry {
        mark_viewed: false,
        manifest,
      },
    );
  }

  pub fn remove(&mut self, file_path: &Vec<u8>) {
    self.files.remove(file_path);
  }

  pub fn mark(&mut self, file_path: &Vec<u8>) {
    if let Some(entry) = self.files.get_mut(file_path) {
      entry.mark_viewed = true;
    }
  }

  pub fn getEntry(&self, file_path: &Vec<u8>) -> Option<&IndexFileEntry> {
    self.files.get(file_path)
  }
}
